### PR TITLE
PAINTROID-274 Refactor JavaCropAlgorithm to Kotlin

### PR DESCRIPTION
**Pull request by CodeChamp-SS**
Refactored JavaCropAlgorithm to Kotlin.
Files that were refactored:

1. CropAlgorithm
2. JavaCropAlgorithm
3. TransformToolIntegrationTest
4. JavaCropAlgorithmTest

https://jira.catrob.at/browse/PAINTROID-274

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Paintroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *#paintroid* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/CropAlgorithm.kt [hunk 1]
```diff
@@ -16,12 +16,11 @@
...
+import android.graphics.Bitmap
+import android.graphics.Rect
 
-import android.graphics.Bitmap;
-import android.graphics.Rect;
-
-public interface CropAlgorithm {
-	Rect crop(Bitmap bitmap);
+interface CropAlgorithm {
+    fun crop(bitmap: Bitmap?): Rect?
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/CropAlgorithm.kt:25]:
```suggestion
    fun crop(bitmap: Bitmap): Rect?
```
Look for usages please

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            for (x in fromX..toX) {
+                if (pixels[y][x] != Color.TRANSPARENT) {
+                    return true
+                }
+            }
+        }
+        return false
+    }
+
+    override fun crop(bitmap: Bitmap?): Rect? {
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:46]: Simplify this function, please.

**Review by gPathpp — CHANGES_REQUESTED** [review]

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/CropAlgorithm.kt (hunk 1 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/CropAlgorithm.kt:25]: I had checked it, the bitmap can be nullable, so `?` should be present.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 2 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:90]: can you please have a look at it now? I tried to simplify it as much as I could :)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    ): Boolean {
+        for (y in fromY..toY) {
+            for (x in fromX..toX) {
+                if (pixels[y][x] != Color.TRANSPARENT) {
+                    return true
+                }
+            }
+        }
+        return false
+    }
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:44]:
```suggestion
    private fun lineContainsOpaquePixel(bounds: Rect, pixels: Array<IntArray>, y) {
       ...
    }
    private fun columnContainsOpaquePixel(bounds: Rect, pixels: Array<IntArray>, x) {
       ...
    }
```
https://kotlinlang.org/docs/collection-elements.html#retrieve-by-condition

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    private fun adjustTop(bounds: Rect, pixels: Array<IntArray>): Int {
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:46]: Please use  https://kotlinlang.org/docs/collection-elements.html#retrieve-by-condition

Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:46]: So this for the other "adjust" methods too, please. Also, rename them please, since they don't adjust anything but determine the most left, right, top, and bottom Opaque pixel :)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    override fun crop(bitmap: Bitmap?): Rect? {
+        if (bitmap == null) {
+            Log.e(TAG, "bitmap is null!")
+            return null
+        }
+        val pixels = Array(bitmap.height) { IntArray(bitmap.width) }
+        for (i in 0 until bitmap.height) {
+            bitmap.getPixels(pixels[i], 0, bitmap.width, 0, i, bitmap.width, 1)
+        }
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:98]: Move this please to an extra function.

**Review by gPathpp — CHANGES_REQUESTED** [review]

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 4 above)
Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:46]: As requested, here is a pseudo code ;)
```suggestion
    fun getColumnUppermostOpaquePixel(a: Matrix) = 
    a.ColRange.firstOrNull{ col -> a.rowRange.firstOrNull{a[col][row] is OPAQUE } != null}
```
For the other methods use [lastOrNull](https://kotlinlang.org/docs/collection-elements.html#retrieve-by-condition) (if needed  ;))
You can write an [extension method](https://kotlinlang.org/docs/extensions.html) for the retrieval of the ranges  ;)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 5 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: done :)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 4 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: thanks! made the suggested changes ;D

**Review by geolas94 — APPROVED** [review]
lgtm

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import android.graphics.Bitmap
+import android.graphics.Color
+import android.graphics.Rect
+import android.util.Log
+
+private const val TAG = "cropAlgorithmSnail"
+
+class JavaCropAlgorithm : CropAlgorithm {
+    private fun getRange(s: Int, e: Int): ArrayList<Int> {
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:29]: Please use [kotlin extenstion functions](https://kotlinlang.org/docs/extensions.html)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        val range = ArrayList<Int>()
+        for (i in s..e) range.add(i)
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:31]: Why? You could simply return s..e.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 6 above)
Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:29]: `fun Rect.rowRange() = this.top..this.bottom`
`fun Rect.colRange() = this.left..this.right`

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            return null
+        }
+        val pixels = initPixels(bitmap)
+        val bounds = Rect(0, 0, bitmap.width - 1, bitmap.height - 1)
+        val top = getTopmostCellWithOpaquePixel(bounds, pixels)
+        if (top == null) {
+            Log.i(TAG, "nothing to crop")
+            return top
+        }
+        bounds.top = top
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:87]:
```suggestion
        bounds.top = getTopmostCellWithOpaquePixel(bounds, pixels) ?: run { 
            Log.i(TAG, "nothing to crop") 
            return null
        }
```

Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:87]: Prev code bit obfuscating, since top is a nullable integer which was returned for a nullable Rect.

Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:87]: Since this seems more like a debug log, you could simply delete it.
```suggestion
        bounds.top = getTopmostCellWithOpaquePixel(bounds, pixels) ?: return null
```

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+package org.catrobat.paintroid.tools.helper
+
+import android.graphics.Bitmap
+import android.graphics.Color
+import android.graphics.Rect
+import android.util.Log
+
+private const val TAG = "cropAlgorithmSnail"
+
+class JavaCropAlgorithm : CropAlgorithm {
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:28]: Please rename.
Also, are there more than one crop algorithm?
Or is this again an interface with only one implementation?

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                bounds.right
+            ).firstOrNull { col -> pixels[row][col] != Color.TRANSPARENT } != null
+        }
+
+    private fun getLeftmostCellWithOpaquePixel(bounds: Rect, pixels: Array<IntArray>): Int? =
+        getRange(bounds.left, bounds.right).firstOrNull { col ->
+            getRange(
+                bounds.top,
+                bounds.bottom
+            ).firstOrNull { row -> pixels[row][col] != Color.TRANSPARENT } != null
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:56]: Instead of creating an Array<Array> we could simply use the bitmap and [getPixel](https://developer.android.com/reference/kotlin/android/graphics/Bitmap#getpixel)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+    private fun initPixels(bitmap: Bitmap): Array<IntArray> {
+        val pixels = Array(bitmap.height) { IntArray(bitmap.width) }
+        for (i in 0 until bitmap.height) {
+            bitmap.getPixels(pixels[i], 0, bitmap.width, 0, i, bitmap.width, 1)
+        }
+        return pixels
+    }
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:73]:
```suggestion
```

**Review by gPathpp — CHANGES_REQUESTED** [review]

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 10 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: done :)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 9 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:28]: Interface with only one implementation.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 8 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: deleted the debug log.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 7 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: removed it, as it was no longer required.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 9 above)
Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:28]: You already know what to do ;)

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import android.graphics.Rect
+import android.util.Log
+
+private const val TAG = "cropAlgorithmSnail"
+
+class JavaCropAlgorithm : CropAlgorithm {
+    private lateinit var bounds: Rect
+    private lateinit var bitmap: Bitmap
+
+    private fun Rect.rowRange() = this.top..this.bottom
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:32]: Sorry, since we use x and y here, please refactor it respectively.
```suggestion
    private fun Rect.yRange() = this.top..this.bottom
```

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                .firstOrNull { y -> isOpaquePixel(x, y) } != null
+        }
+
+    override fun crop(bitmap: Bitmap?): Rect? {
+        if (bitmap == null) {
+            Log.e(TAG, "bitmap is null!")
+            return null
+        }
+
+        this.bitmap = bitmap
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:68]:
```suggestion
        this.bitmap = bitmap?:return null
```

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 12 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:32]: hmm.. ok

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 9 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:28]: what? 😅😂

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        bounds = Rect(0, 0, bitmap.width - 1, bitmap.height - 1)
+
+        val top = getTopmostCellWithOpaquePixel() ?: return null
+        bounds.top = top
+        bounds.bottom = getBottommostCellWithOpaquePixel() ?: bounds.top - 1
+        bounds.left = getLeftmostCellWithOpaquePixel() ?: bounds.right + 1
+        bounds.right = getRightmostCellWithOpaquePixel() ?: bounds.left - 1
```
Comment (gPathpp) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:75]: The right side of each ?: won't be executed, and if there would be a bug in the algorithm.
Anyway, I think it's ok :+1:

File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt [hunk 15]
```diff
@@ -0,0 +1,1367 @@
...
+                0,
+                10,
+                verticalStartX,
+                verticalStartY,
+                10,
+                lineLength
+            )
+        }
+    }
+}
```
Comment (gPathpp) [inline Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt:1367]: This test looks worse in kotlin. Please revert this change and make an extra ticket in which improvements to the test are then applied.

**Review by gPathpp — CHANGES_REQUESTED** [review]

File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt (hunk 15 above)
Comment (CodeChamp-SS) [reply Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt:1367]: The test would fail otherwise 😅

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 12 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: done.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 13 above)
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt]: done.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt (hunk 9 above)
Comment (gPathpp) [reply Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaCropAlgorithm.kt:25]: Since there will only be one crop algorithm, you can delete the interface and rename the implementation to CropAlgorithm

File: Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt (hunk 15 above)
Comment (gPathpp) [reply Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt:1367]: Can't think of a reason why it would fail. :shrug: 
Please make the necessary change in java, the kotlin test code is unreadable.

Comment (CodeChamp-SS) [reply Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt:1367]: runBlocking won't be possible in Java

Comment (gPathpp) [reply Paintroid/src/androidTest/java/org/catrobat/paintroid/test/espresso/tools/TransformToolIntegrationTest.kt:1367]: Hmm, that's unfortunate... Can you extract the affected test cases into a file? (Since this test file is anyway too big)

**Review by gPathpp — APPROVED** [review]
